tests/workflows/test_snowflake.py

```diff
@@ -0,0 +1,172 @@
+    engine = create_engine(URL(**connection_kwargs))
+    engine.execute(f"DROP TABLE IF EXISTS {name}")
+    engine.execute(
+        f"""create table if not exists {name} (
+                ride_id varchar not null unique,
+                rideable_type varchar not null,
+                started_at timestamp not null,
+                ended_at timestamp not null,
+                start_station_name varchar not null,
+                start_station_id smallint not null,
+                end_station_name varchar not null,
+                end_station_id smallint not null,
+                start_lat number,
+                start_lng number,
+                end_lat number,
+                end_lng number,
+                is_member boolean not null
+            )"""
+    )
+    yield name
+    # after the data is written, delete table
+    engine.execute(f"DROP TABLE IF EXISTS {name}")
+
+
+@pytest.mark.client("snowflake")
+def test_write(client, connection_kwargs, table):
+    csv_path = "nyc-bike-data"
+
+    def unzip(filename: str):
+        """Unzip csv file to csv_path"""
+        if not os.path.exists(f"{csv_path}/{filename}"):
+            url = f"https://s3.amazonaws.com/tripdata/{filename}.zip"
+            try:
+                zip_path, _ = urllib.request.urlretrieve(url)
+            except urllib.error.HTTPError:
+                # some of the filenames have "citbike" vs "citibike"
+                zip_path, _ = urllib.request.urlretrieve(
+                    url.replace("-citibike-", "-citbike-")
+                )
+            with zipfile.ZipFile(zip_path, "r") as f:
+                f.extractall(csv_path)
+        return filename
+
+    filenames = [
+        f"{ts.year}{ts.month:02}-citibike-tripdata.csv"
+        for ts in pd.date_range(start="2022-01-01", end="2023-03-01", freq="MS")
+    ]
+
+    # download and unzip files
+    futures = client.map(unzip, filenames)
+    wait(futures)
+
+    # preprocess data
+    def safe_int(x):
+        """Some station IDs are not correct integers"""
+        try:
+            return int(float(x))
+        except Exception:
+            # if station ID is not an int, return -1
+            return -1
+
+    ddf = dd.read_csv(
+        f"{csv_path}/*.csv",
+        converters={"start_station_id": safe_int, "end_station_id": safe_int},
+    )
+
+    # filter out incorrect station IDs
+    ddf = ddf[(ddf.start_station_id != -1) & (ddf.end_station_id != -1)].reset_index(
+        drop=True
+    )
+
+    # create boolean is_member and drop member_casual
+    ddf["is_member"] = ddf.member_casual == "member"
+
+    ddf = ddf.drop(columns="member_casual")
+
+    # repartition to ensure even chunks
+    ddf = ddf.repartition(partition_size="100Mb")
+
+    wait(ddf)
+
+    # save data to Snowflake
+    to_snowflake(ddf, name=table, connection_kwargs=connection_kwargs)
+
+
+@pytest.mark.client("snowflake")
+def test_read_explore(client, connection_kwargs, table):
```

Comment:
This workflow feels like something that I personally would solve in Snowflake directly.

Is there a way we can make this more iterative/explorative so that we'd benefit from having intermediate results available on the cluster? Alternatively, could we make more use of different Dask libraries/integrations? This might look more similar to https://github.com/Snowflake-Labs/snowpark-python-demos.

Reply:
I'm not sure what the exact motivation of this is, maybe we could also go the other way and dumb it down until it becomes a narrow test of Snowflake read performance.

Reply:
> Is there a way we can make this more iterative/explorative so that we'd benefit from having intermediate results available on the cluster

That't why I added quantiles and filtering out outliers.

> I'm not sure what the exact motivation of this is, maybe we could also go the other way and dumb it down until it becomes a narrow test of Snowflake read performance

I'm a little struggling with this myself. It seems that everyone has a different idea about the purpose of this, and how complicated / representative of a real-world it should look. This is an attempt to find a balance between "very dumb read and calculate a single mean or sum" and "read and do something very complicated, like train an ML model, or perform a full-on ETL pipeline".

Maybe @jrbourbeau and @mrocklin can advise?